src/main/java/blusunrize/immersiveengineering/common/items/BuzzsawItem.java

```diff
@@ -0,0 +1,775 @@
+				if(!foundLog)
+					break;
+			}
+		}
+		return logs >= 3&&leafTop;
+	}
+
+	private boolean canFellTree(ItemStack stack)
+	{
+		ItemStack sawblade = getSawblade(stack);
+		if(sawblade.getItem() instanceof SawbladeItem)
+			return ((SawbladeItem)sawblade.getItem()).canSawbladeFellTree();
+		return false;
+	}
+
+	/**
+	 * The max distance a block can be from the initial hit
+	 * to still be considered part of the tree
+	 * This is based on the largest vanilla Jungle Trees
+	 */
+	private static final int MAX_HORIZONTAL_DISTANCE = 7;
+
+	private boolean fellTree(World world, BlockPos initialPos, ServerPlayerEntity player, ItemStack stack)
+	{
+		int logs = 0;
+		List<BlockPos> openList = new ArrayList<>();
+		List<BlockPos> closedList = new ArrayList<>();
+		openList.add(initialPos);
+		while(!openList.isEmpty()&&closedList.size() < 512&&logs < 256)
+		{
+			BlockPos next = openList.remove(0);
+
+			// Ignore blocks too far away
+			if(Math.abs(next.getX()-initialPos.getX()) > MAX_HORIZONTAL_DISTANCE
+					||Math.abs(next.getZ()-initialPos.getZ()) > MAX_HORIZONTAL_DISTANCE)
+				continue;
+
+			if(!closedList.contains(next))
+			{
+				BlockState state = world.getBlockState(next);
+				if(state.isIn(BlockTags.LOGS))
+				{
+					closedList.add(next);
+					logs++;
+					// Find all at same level or above, including diagonals
+					for(int y = 0; y <= 1; y++)
+						for(int z = -1; z <= 1; z++)
+							for(int x = -1; x <= 1; x++)
+								openList.add(next.add(x, y, z));
+				}
+				else if(state.isIn(BlockTags.LEAVES))
+				{
+					closedList.add(next);
+					int trunkDist = state.getBlock() instanceof LeavesBlock?state.get(LeavesBlock.DISTANCE): 0;
+					// Leaves only propagate in cardinal directions, and only to other leaves
+					for(Direction dir : new Direction[]{Direction.UP, Direction.NORTH, Direction.SOUTH, Direction.WEST, Direction.EAST})
+					{
+						BlockPos adj = next.offset(dir);
+						BlockState adjState = world.getBlockState(adj);
+						if(adjState.isIn(BlockTags.LEAVES))
+						{
+							int adjDist = adjState.getBlock() instanceof LeavesBlock?adjState.get(LeavesBlock.DISTANCE): 0;
+							if(adjDist < trunkDist) // We don't want to get closer
+								continue;
+						}
+						openList.add(adj);
+					}
+				}
+			}
+		}
+
+		if(closedList.size()==0)
+			return false;
+		// Register a Tick Handler to break the blocks, 10 at a time
+		MinecraftForge.EVENT_BUS.register(new Object()
+		{
+			@SubscribeEvent
+			public void onTick(TickEvent.WorldTickEvent event)
+			{
+				breakFromList(closedList, 5, world, player, stack);
+				if(closedList.isEmpty())
+					MinecraftForge.EVENT_BUS.unregister(this);
+			}
+		});
+		return true;
+	}
+
+	private void breakFromList(List<BlockPos> closedList, int maxAmount, World world, ServerPlayerEntity player, ItemStack stack)
+	{
+		int count = 0;
+		while(count++ < maxAmount&&!closedList.isEmpty())
+		{
+			BlockPos pos = closedList.remove(0);
```

Comment:
Again, either remove from the end or use a data structure where the first element can be removed efficiently.